### PR TITLE
Implement when by cribbing from if and begin.

### PR DESCRIPTION
**Pull request by technomancy**
I've got some cases where it would be very handy to use `when` instead of `if`+`begin`.

I took a stab at implementing it, but when I go to try it out on a device, the body seems to be skipped every time.

Do you think you could take a look and see if I'm making any obvious mistakes? I've tried combining elements from the codegen implementations of `Branch` and `Sequence` but something isn't right.

**Edit**: I found the problem; I misunderstood what `CPSE` was. My latest push seems to fix this, but the implementation is a bit tacky; it uses two `CPSE` instructions to simulate "Compare Skip Not Equal"; I'm sure there's a better way to do it.

### CODE REVIEW

File: src/codegen.c [hunk 1]
```diff
@@ -81,6 +81,17 @@ void codegen_emit(AST_expr *expr, int parent_numArgs, FILE *outputFile) {
 			fprintf(outputFile, "if_end%i:\n", label1);
 			break;
 
+        case When:
+			label1 = if_end_unique++;
+			codegen_emit(expr->body[0], parent_numArgs, outputFile);
+			fprintf(outputFile, "\tCPSE CRSh, falseReg\n\tCPSE falseReg, falseReg\n");
```
Comment (ryansuchocki) [inline src/codegen.c]: CPSE is "ComPare, Skip (next instruction) if Equal". The JMP says "jump over all of the following instructions until the if_end label". The CRS register holds the result of the previous expression (which in this case is the 'when' condition). Therefore, these two instructions say "if the condition was equal to false then **don't** jump over these body expressions".

I don't think there's an inverse of CPSE, so do something like "CP CRSh, falseReg; BREQ if_end", which is "compare the result with false and branch to if_end if they were equal."

Comment (technomancy) [reply src/codegen.c]: Perfect; thanks! I knew there was something simpler. I'll see if I can spend a little time reading up on AVR before doing any more patches like this; do you have a resource you would recommend?

Comment (ryansuchocki) [reply src/codegen.c]: I think the Atmel datasheets are pretty good, and [this](http://atmel-studio-doc.s3-website-us-east-1.amazonaws.com/webhelp/GUID-0B644D8F-67E7-49E6-82C9-1B2B9ABE6A0D-en-US-1/index.html) is my go-to reference for the instruction set. I can't think of any other resources in particular. When I was starting out I found it very useful to write simple C programs and compile to assembly (`avr-gcc -S`) or compile fully then disassemble with `avr-objdump` to inspect the result.

**Comment by ryansuchocki** [comment]
@technomancy are you happy that it works as intended?

**Comment by technomancy** [comment]
Thanks, yes; with the change you suggested this seems to do the trick.

**Comment by technomancy** [comment]
Thanks!

**Pull request merged**